dex/txfee/oracle.go

```diff
@@ -0,0 +1,255 @@
+package txfee
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"decred.org/dcrdex/dex"
+)
+
+// Oracle provides transaction fees for all configured assets from external
+// sources. Fee estimate values are in atoms for dcr, gwei for ethereum,
+// satoshis for bitcoin and bitcoin clone blockchains (per byte sat), or the
+// lowest non-divisible unit in other non-Bitcoin blockchains.
+type Oracle struct {
+	chainIDs       []uint32
+	sources        []*feeEstimateSource
+	feeMtx         sync.RWMutex
+	txFeeEstimates map[uint32]*Estimate
+
+	listenersMtx sync.RWMutex
+	listeners    map[string]chan<- map[uint32]*Estimate
+}
+
+// NewOracle returns a new instance of *Oracle.
+func NewOracle(net dex.Network, cfg Config, chainsIDs []uint32) (*Oracle, error) {
+	if len(chainsIDs) == 0 {
+		return nil, errors.New("provide chainIDs to fetch fee estimate for")
+	}
+
+	if net != dex.Mainnet && net != dex.Testnet {
+		return nil, errors.New("fee estimate oracle is available for only mainnet and testnet")
+	}
+
+	o := &Oracle{
+		chainIDs:       chainsIDs,
+		sources:        feeEstimateSources(net, cfg),
+		txFeeEstimates: make(map[uint32]*Estimate),
+		listeners:      map[string]chan<- map[uint32]*Estimate{},
+	}
+
+	for _, chainID := range chainsIDs {
+		if sym := dex.BipIDSymbol(chainID); sym == "" {
+			return nil, fmt.Errorf("chainID %d is invalid", chainID)
+		}
+
+		// Init chain.
+		o.txFeeEstimates[chainID] = new(Estimate)
+	}
+
+	return o, nil
+}
+
+// FeeEstimates retrieves the current fee estimates.
+func (o *Oracle) FeeEstimates() map[uint32]*Estimate {
+	o.feeMtx.RLock()
+	defer o.feeMtx.RUnlock()
+	feeEstimates := make(map[uint32]*Estimate, len(o.txFeeEstimates))
+	for chainID, feeEstimate := range o.txFeeEstimates {
+		if feeEstimate.Value > 0 && time.Since(feeEstimate.LastUpdated) < FeeEstimateExpiry {
+			fe := *feeEstimate
+			feeEstimates[chainID] = &fe
+		}
+	}
+	return feeEstimates
+}
+
+// AddFeeListener adds a new tx fee listener. If the uniqueID already exists, it
+// will be overridden by the new feeChan. feeChan MUST never be closed by
+// callers, use RemoveFeeListener instead.
+func (o *Oracle) AddFeeListener(uniqueID string, txFeeChan chan<- map[uint32]*Estimate) {
+	o.listenersMtx.Lock()
+	defer o.listenersMtx.Unlock()
+	o.listeners[uniqueID] = txFeeChan
+}
```

Comment:
Yeah, tatanka won't be the only consumer of this pkg.

Reply:
But would you ever have two consumers at once?

Reply:
Wallets could just query the rates as needed, they don't need live updates.

Reply:
> Wallets could just query the rates as needed, they don't need live updates.

In fact, I'd even go so far as to say that wallets should be the primary source of fee information, though I recognize that presents challenges because sources provide fees for more that one asset at a time. How you have it here is good enough start. It'll change a lot before mesh is ready.